crates/bevy_render/src/picking.rs

```diff
@@ -0,0 +1,434 @@
+            // This is to empty the channel before we continue
+            rx.try_recv().expect("map_async should have been completed");
+        }
+
+        let send_index = current_buffer_index.0;
+        let buffer_slice = buffers.entity_buffers[send_index].slice(..);
+        let buffer_view = buffer_slice.get_mapped_range();
+        let mesh_id_buffer = buffer_view.to_vec();
+        // We have to make sure all mapped views are dropped before we unmap the buffer.
+        drop(buffer_view);
+        // We need to unmap the buffer because it will be used in the next frame and can't be mapped at that point
+        buffers.entity_buffers[send_index].unmap();
+
+        // Send the data to the main world
+        if let Err(err) = gpu_picking_camera.sender.try_send(GpuPickingData {
+            padded_bytes_per_row: buffers.buffer_dimensions.padded_bytes_per_row,
+            mesh_id_buffer,
+            visible_mesh_entities: visible_mesh_entities.clone(),
+        }) {
+            bevy_log::error!("Failed to send entity buffer: {err}");
+        }
+    }
+}
+
+/// Receives the mesh id buffer from the render world
+fn receive_buffer(mut cameras: Query<&mut GpuPickingCamera>) {
+    for mut cam in &mut cameras {
+        let (_, receiver) = cam.data_channel.clone();
+        let Ok(data) = receiver.try_recv() else { continue; };
+        cam.data = data;
+    }
+}
+
+/// The textures used to draw the entity for each rendered mesh
+#[derive(Component, Clone)]
+pub struct VisibleMeshIdTextures {
+    pub main: CachedTexture,
+    pub sampled: Option<CachedTexture>,
+}
+
+impl VisibleMeshIdTextures {
+    /// This is the color that will represent "no entity" in the mesh id buffer
+    pub fn clear_color() -> wgpu::Color {
+        Color {
+            r: 0.0,
+            g: 0.0,
+            b: 0.0,
+            a: 0.0,
+        }
+    }
+
+    /// Retrieve this target's color attachment. This will use [`Self::sampled`] and resolve to [`Self::main`] if
+    /// the target has sampling enabled. Otherwise it will use [`Self::main`] directly.
+    pub fn get_color_attachment(&self, ops: Operations<wgpu::Color>) -> RenderPassColorAttachment {
+        match &self.sampled {
+            Some(sampled_texture) => RenderPassColorAttachment {
+                view: &sampled_texture.default_view,
+                resolve_target: Some(&self.main.default_view),
+                ops,
+            },
+            None => RenderPassColorAttachment {
+                view: &self.main.default_view,
+                resolve_target: None,
+                ops,
+            },
+        }
+    }
+}
+
+/// This creates the required buffers for each camera
+fn prepare_buffers(
+    render_device: Res<RenderDevice>,
+    mut cameras: Query<
+        (Entity, &ExtractedCamera, &mut ExtractedGpuPickingCamera),
+        Changed<ExtractedCamera>,
+    >,
+    mut buffer_cache: Local<HashMap<Entity, (BufferDimensions, [Buffer; BUFFER_COUNT])>>,
```

Comment:
Blocking: Memory leak. In a situation where new cameras are spawned periodically, (expected, supposing that the user uses scene cameras and switches scenes periodically) this will occupy an increasing amount of memory without ever releasing any.

A possibility is to cap the number of entries in the cache, removing the older ones, with a `VecDequeu<Entity>` or `BinaryHeap` to keep track of entity age.

Reply:
Cleaning up makes much more sense.

Reply:
Done